prog/dftb+/prg_dftb/main.F90

```diff
@@ -1329,17 +1323,72 @@ subroutine handleCoordinateChange(coord0, latVec, invLatVec, species0, mCutoff,
     if (allocated(sccCalc)) then
       call sccCalc%updateCoords(coord, species, neighborList, img2CentCell)
     end if
-    if (allocated(dispersion)) then
-      call dispersion%updateCoords(neighborList, img2CentCell, coord, &
-          & species0)
-    end if
     if (allocated(thirdOrd)) then
       call thirdOrd%updateCoords(neighborList, species)
     end if
 
   end subroutine handleCoordinateChange
 
+  !> Does the operations that are necessary after an SCC cycle, even if there is only one iteration
+  !> (i.e. non-SCC)
+  subroutine handlePostSccChanges(tLatticeChanged, nAtom, latVecs, species, mCutoff, dispersion, &
+      & pRepCont, img2CentCell, nNeighbor, neighborList, coord, energy)
+
+    !> Is there a  need to update lattice vectors (both periodic and they have changed)
+    logical, intent(in) :: tLatticeChanged
+
+    !> Number of actual (central cell) atoms
+    integer, intent(in) :: nAtom
+    
+    !> Lattice vectors if periodic
+    real(dp), intent(inout) :: latVecs(:,:)
+
+    !> chemical species of all atoms
+    integer, intent(inout) :: species(:)
+
+    !> Longest cutoff distance that neighbour maps are generated
+    real(dp), intent(inout) :: mCutoff
+
+    !> Dispersion interactions
+    class(DispersionIface), allocatable, intent(inout) :: dispersion
 
+    !> Repulsive interaction data
+    type(ORepCont), intent(in) :: pRepCont
+    
+    !> image atoms to their equivalent in the central cell
+    integer, allocatable, intent(inout) :: img2CentCell(:)
+
+    !> Number of neighbours for each actual atom
+    integer, intent(in) :: nNeighbor(:)
+    
+    !> List of neighbouring atoms
+    type(TNeighborList), intent(inout) :: neighborList
+
+    !> Coordinates of all atoms including images
+    real(dp), allocatable, intent(inout) :: coord(:,:)
+
+    !> Energy contributions
+    type(TEnergies), intent(inout) :: energy
+    
+    ! Notify various relevant modules about current coordinates/boundary conditions
+
+    call calcRepulsiveEnergy(coord, species, img2CentCell, nNeighbor, neighborList, pRepCont,&
+        & energy%atomRep, energy%ERep)
+    
+    if (allocated(dispersion)) then
+      if (tLatticeChanged) then
+        call dispersion%updateLatVecs(latVecs)
+        mCutoff = max(mCutoff, dispersion%getRCutoff())
```

Comment:
I think, that won't work robustly enough. The neighbour list for the current geometry has been already created at that point. If (although very unlikely) the dispersion needed a longer real space cutoff as all other modules, it would not contain enough neighbors for the following `updateCoords()` call. We should probaly make the `upateLatVecs()` still at the beginning of the geoopt-loop.